org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java

```diff
@@ -234,27 +195,37 @@ public static String getPackageName(IJavaProject javaProject, URI uri) {
 			File file = ResourceUtils.toFile(uri);
 			//FIXME need to determine actual charset from file
 			String content = Files.toString(file, Charsets.UTF_8);
-			if (content.isEmpty() && javaProject != null && ProjectsManager.DEFAULT_PROJECT_NAME.equals(javaProject.getProject().getName())) {
-				java.nio.file.Path path = Paths.get(uri);
-				java.nio.file.Path parent = path;
-				while (parent.getParent() != null && parent.getParent().getNameCount() > 0) {
-					parent = parent.getParent();
-					String name = parent.getName(parent.getNameCount() - 1).toString();
-					if (SRC.equals(name)) {
-						String pathStr = path.getParent().toString();
-						if (pathStr.length() > parent.toString().length()) {
-							pathStr = pathStr.substring(parent.toString().length() + 1);
-							pathStr = pathStr.replace(PATH_SEPARATOR, PERIOD);
-							return pathStr;
-						}
-					}
-				}
+			if (content.isEmpty() && javaProject != null && !JDTStandaloneFileUtils.isVisibleProject(javaProject.getProject())) {
+				return guessPackageNameFromPath(uri);
 			} else {
 				return getPackageName(javaProject, content);
 			}
 		} catch (IOException e) {
-			JavaLanguageServerPlugin.logException("Failed to read package name from "+uri, e);
+			JavaLanguageServerPlugin.logException("Failed to read package name from " + uri, e);
 		}
+
+		return "";
+	}
+
+	private static String guessPackageNameFromPath(URI uri) {
+		java.nio.file.Path javaPath = Paths.get(uri);
+		IPath containerPath = ResourceUtils.filePathFromURI(javaPath.getParent().toUri().toString());
+		IPath workspaceRoot = JDTStandaloneFileUtils.getWorkspaceRoot(containerPath);
+		if (workspaceRoot == null) {
+			List<String> segments = Arrays.asList(containerPath.segments());
+			if (segments.contains(SRC)) {

```

Comment:
should be first process src/main/java and  src/test/java

Reply:
fixed